R/driver-snowflake.R

```diff
@@ -87,3 +87,212 @@ setMethod("odbcDataType", "Snowflake",
+#' detects ambient OAuth credentials on platforms like Snowpark Container
+#' Services or Posit Workbench.
+#'
+#' @inheritParams DBI::dbConnect
+#' @param account A Snowflake [account
+#'   identifier](https://docs.snowflake.com/en/user-guide/admin-account-identifier),
+#'   e.g. `"testorg-test_account"`.
+#' @param driver The name of the Snowflake ODBC driver, or `NULL` to use the
+#'   default name.
+#' @param warehouse The name of a Snowflake compute warehouse, or `NULL` to use
+#'   the default.
+#' @param database The name of a Snowflake database, or `NULL` to use the
+#'   default.
+#' @param schema The name of a Snowflake database schema, or `NULL` to use the
+#'   default.
+#' @param uid,pwd Manually specify a username and password for authentication.
+#'   Specifying these options will disable ambient credential discovery.
+#' @param ... Further arguments passed on to [`dbConnect()`].
+#'
+#' @returns An `OdbcConnection` object with an active connection to a Snowflake
+#'   account.
+#'
+#' @examples
+#' \dontrun{
+#' # Use ambient credentials.
+#' DBI::dbConnect(odbc::snowflake())
+#'
+#' # Use browser-based SSO (if configured). Only works on desktop.
+#' DBI::dbConnect(
+#'   odbc::snowflake(),
+#'   account = "testorg-test_account",
+#'   authenticator = "externalbrowser"
+#' )
+#'
+#' # Use a traditional username & password.
+#' DBI::dbConnect(
+#'   odbc::snowflake(),
+#'   account = "testorg-test_account",
+#'   uid = "me",
+#'   pwd = rstudioapi::askForPassword()
+#' )
+#' }
+#' @export
+snowflake <- function() {
+  new("Snowflake")
+}
+
+#' @rdname snowflake
+#' @export
+setMethod(
+  "dbConnect", "Snowflake",
+  function(drv,
+           account = Sys.getenv("SNOWFLAKE_ACCOUNT"),
+           driver = NULL,
+           warehouse = NULL,
+           database = NULL,
+           schema = NULL,
+           uid = NULL,
+           pwd = NULL,
+           ...) {
+    args <- snowflake_args(
+      account = account,
+      driver = driver,
+      warehouse = warehouse,
+      database = database,
+      schema = schema,
+      uid = uid,
+      pwd = pwd,
+      ...
+    )
+    inject(dbConnect(odbc(), !!!args))
+  }
+)
+
+snowflake_args <- function(account = Sys.getenv("SNOWFLAKE_ACCOUNT"),
+                           driver = NULL,
+                           ...) {
+  args <- list(
+    driver = driver %||% snowflake_default_driver(),
+    account = account,
+    server = snowflake_server(account),
+    # Connections to Snowflake are always over HTTPS.
+    port = 443
+  )
+  auth <- snowflake_auth_args(account, ...)
+  all <- utils::modifyList(c(args, auth), list(...))
+
+  # Respect the Snowflake Partner environment variable, if present.
+  if (is.null(all$application) && nchar(Sys.getenv("SF_PARTNER")) != 0) {
+    all$application <- Sys.getenv("SF_PARTNER")
+  }
+
+  arg_names <- tolower(names(all))
+  if (!"authenticator" %in% arg_names && !all(c("uid", "pwd") %in% arg_names)) {
+    warn(
```

Comment:
Should this be a warning or an error?

Reply:
I'm emulating the `odbc::databricks()` call here, which just issues a warning. I'm open to changing both, though.

Reply:
I think we should make it an error for both. If you don't want to do in this PR, feel free to file an issue.